### PR TITLE
Tests for DNS Link functions

### PR DESCRIPTION
**Pull request by jcoglan**
This is James from @neighbourhoodie, I have been working on the STF milestone M1 to improve test coverage, focussing on `resolved`. We have written a lot of tests for this module and we are splitting the work into a number of PRs, one for each test file we've added. Each new test file usually corresponds to a single source file, and exercises the functions in that source file directly.

This PR includes tests for `resolved-link.c`, providing coverage of functions relating to `Link` objects.

### CODE REVIEW

**Comment by jcoglan** [comment]
I believe I have fixed all the issues with the files in this PR -- asan/ubsan problems, gcc/clang differences, linter checks, etc. and the remaining failing CI tasks are failing for reasons other than the changes here. If that's incorrect and there are other things I still need to change, do let me know what else I need to do.

File: src/resolve/test-resolved-link.c [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+        LinkAddress *address;
+        DnsServerType server_type;
+        union in_addr_union server_addr;
+        char *server_name;
+        uint16_t server_port;
+        DnsServer *server;
+} LinkAllocEnv;
+
+static void link_alloc_env_teardown(LinkAllocEnv *env) {
+        free(env->server_name);
```
Comment (bluca) [inline src/resolve/test-resolved-link.c:150]: either assert(env) or check if null and exit early

File: src/resolve/test-resolved-link.c [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+        link_address_free(env->address);
+        dns_server_unref(env->server);
+        sd_event_unref(env->manager.event);
+}
+
+static void link_alloc_env_setup(LinkAllocEnv *env, int family, DnsServerType server_type) {
+        Link *link = NULL;
+
```
Comment (bluca) [inline src/resolve/test-resolved-link.c:158]: assert(env)

File: src/resolve/test-resolved-link.c [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+        env->manager = (Manager) {};
+        env->ifindex = 1;
+
+        sd_event_new(&env->manager.event);
```
Comment (bluca) [inline src/resolve/test-resolved-link.c:164]: this needs an assert too

**Comment by jcoglan** [comment]
@bluca I've added the requested null checks, as well as after anything else that allocates.

**Review by bluca — APPROVED** [review]

**Pull request merged**